etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/database/PostgresDao.java

```diff
@@ -137,6 +142,91 @@ public PartnerMetadata fetchMetadata(String submissionId)
         }
     }
 
+    @Override
+    public Set<PartnerMetadata> fetchMetadataForMessageLinking(String submissionId)
+            throws SQLException, FormatterProcessingException {
+        var sql =
+                """
+                SELECT m2.*
+                FROM metadata m1
+                JOIN metadata m2
+                    ON m1.placer_order_number = m2.placer_order_number
+                        AND m1.sending_application_id = m2.sending_application_id
+                        AND m1.sending_facility_id = m2.sending_facility_id
+                WHERE m1.sent_message_id = ?;
+                """;
+
+        try (Connection conn = connectionPool.getConnection();
+                PreparedStatement statement = conn.prepareStatement(sql)) {
+            statement.setString(1, submissionId);
+            ResultSet resultSet = statement.executeQuery();
+
+            Set<PartnerMetadata> metadataSet = new HashSet<>();
+
+            while (resultSet.next()) {
+                metadataSet.add(partnerMetadataFromResultSet(resultSet));
+            }
+
+            return metadataSet;
+        }
+    }
+
+    @Override
+    public Optional<MessageLink> fetchMessageLink(String messageId) throws SQLException {
+        var sql =
+                """
+                SELECT *
+                FROM message_link
+                WHERE message_id = ?;
+                """;
+
+        UUID linkId = null;
+        Set<String> messageIds = new HashSet<>();
+        try (Connection conn = connectionPool.getConnection();
+                PreparedStatement statement = conn.prepareStatement(sql)) {
+            statement.setString(1, messageId);
+
+            try (ResultSet resultSet = statement.executeQuery()) {
+                while (resultSet.next()) {
+                    if (linkId == null) {
+                        linkId = UUID.fromString(resultSet.getString("link_id"));
+                    }
+                    messageIds.add(resultSet.getString("message_id"));
+                }
+            }
+        }
+
+        if (!messageIds.isEmpty()) {
+            return Optional.of(new MessageLink(linkId, messageIds));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public void insertMessageLink(MessageLink messageLink) throws SQLException {
+        var sql =
+                """
+                INSERT INTO message_link (link_id, message_id)
+                SELECT ?, ?
+                WHERE NOT EXISTS (
+                    SELECT 1 FROM message_link WHERE message_id = ?
+                );
+                """;
+
+        try (Connection conn = connectionPool.getConnection();
+                PreparedStatement statement = conn.prepareStatement(sql)) {
+            UUID linkId = Optional.ofNullable(messageLink.getLinkId()).orElse(UUID.randomUUID());
```

Comment:
What do you think about moving this logic to the `PartnerMetadataOrchistrator`?

E.g.
`messageLinkStorage.saveMessageLink(new MessageLink(null, messageIds));` is updated to be `messageLinkStorage.saveMessageLink(new MessageLink(UUID.randomUUID(), messageIds));`

Reply:
I'll take a look

Reply:
I agree. After calling `getMessageLink` from the `messageLinkStorage`, the orchestrator can follow up with this logic before calling `saveMessageLink` from the`messageLinkStorage`.

Reply:
Makes sense. Updated it